### ISSUE TITLE
SignOutBBCiD is never used and does not work

### ISSUE THREAD

**Issue by CaptainTK** [vonH/plugin.video.iplayerwww#72]
Although the code has a function to sign out from the BBCiD, it is never used.

@ihurst can you please explain when you originally planned to call SignOutBBCiD?

**Comment by primaeval** [comment]
I'd leave it in even if it is not used right now. It might be useful for switching users without having to switch profiles: eg kids/adults.

**Comment by CaptainTK** [comment]
I am not suggesting to bin it, I want to make sure users can actually log out. Currently, the only way to ever log out is by deleting the cookie file.

**Comment by primaeval** [comment]
Then it should really be used when you disable BBCiD: logout and delete cookies.

**Comment by CaptainTK** [comment]
@ihurst can you also please explain how SignOutBBCiD is supposed to work? Just calling the URL does not appear to clear the login data correctly.

**Comment by primaeval** [comment]
@ihurst It would be nice if you could add to BBCiD Favourites from the addon, if you have time to look at it.

**Comment by CaptainTK** [comment]
I believe this is due to the cookie handling in OpenURL and OpenURLPost. Both contain the line
`cookie_jar.save(ignore_discard=True, ignore_expires=True)`
which means that session cookies are stored and the expiration date is ignored.

If both options are removed, users need to log-in every time they want to access a page which requires a log-in. That is not useful.

If only ignore_expires is removed, there seems to be no difference. SignOutBBCiD still does not work and the log-in is permanently stored.

**Comment by ihurst** [comment]
The session cookies for the iPlayer BBCid are set to expire "at end of session" and I think the issue we have is that we don't have a "session" in the python http client. Each http request is independent, hence if we don't set "ignore_discard=True" then the cookie gets deleted straight away.

All this is a bit of guess work from looking at cookie data and how this system functioned.

The actual cookies that i believe to be relevant for BBCid are called:
- IDENTITY
- IDENTITY-HTTP

The cookie_jar is saved in this file /storage/.kodi/userdata/addon_data/plugin.video.iplayerwww/iplayer.cookies: 

Set-Cookie3: IDENTITY="<data>"; path="/"; domain=".bbc.co.uk"; path_spec; domain_dot; **discard**; version=0

Set-Cookie3: IDENTITY-HTTPS=<data>; path="/"; domain=".bbc.co.uk"; path_spec; domain_dot; secure; **discard;** HttpOnly=None; version=0

We may be able to clear all these cookies using  CookieJar.clear_session_cookies().

I'll try and check this tonight, but I need to get dev environment set back up. If anyone else can try this quickly?

**Comment by CaptainTK** [comment]
Thanks, but I think I already sorted this out in 53979b5197e126f6cda43c9c687888d6bcdc62a9.

The idea is basically the same: clear the cookies in SignOutBBCiD.

**Comment by ihurst** [comment]
Sorry just re-looked at your change and seen that you changed the ignore_discard property on the SignOut request. 

At some point I'd prefer to tidy this up with something like below, but this would require some further changes to the code due to class level variable cookie_jar:

```python
def SignOutBBCiD():
     sign_out_url="https://ssl.bbc.co.uk/id/signout"
     OpenURL(sign_out_url)
     cookie_jar.clear_session_cookies()
```
I've just got my environment set-up so please shout if there is anything I can help with.

**Comment by CaptainTK** [comment]
I tried that first and couldn't get it to work. If you have an idea how to do it properly, please help yourself.

**Comment by ihurst** [comment]
Done and tested a tidy-up for logout.

Also seen there are some other areas that could be tidied i.e. Radio functions that appear in both ipwww_radio and ipwww_video. And we should really be defining the ADDON variable in each file (I'm just seeing issues highlighted by Eclipse (my chosed IDE).

All the new work looks great by the way, you have all been busy!!!

Please shout if there is anything I can help with, sorry I've been away for so long.

**Comment by CaptainTK** [comment]
Thanks for the feedback. I wasn't aware that passing a variable may cause an issue. I will look into the ADDON variable then.

About duplicate functions: A lot of the functions in ipwww_radio and ipwww_video have the same name because they serve the same purpose in their respective scope. However, they are internally completely different and should be kept separate. If you came accross any real duplicates, please give me some pointers and I'll tidy them up.

**Comment by ihurst** [comment]
Function in Video library:

```python
def RadioAddAvailableStreamItem
```

Don't think ADDON is actually causing issue. Just a risk of error when coding.

I really like the re architecture of libs. Previously I looked at separating other functionality into classes. I'll send examples tonight. 

You trying to release soon?

**Comment by CaptainTK** [comment]
Thanks, I must have overlooked this while breaking up the code. Will take care of it tonight.

Yes, I would like to release by the end of the week. I'll have one more week for bugfix releases if required and then I'll be busy with other stuff for 6 weeks or so.

**Comment by CaptainTK** [comment]
Fixed by b46b98317863e6773ccb57065b58c72b39afa30e
